platform-sdk/docs/proposals/metric-labels/README.md

```diff
@@ -0,0 +1,325 @@
+
+## Design & Architecture
+
+This chapter describes the changes that are needed to add labels to the metrics module.
+
+### About labels
+
+Before we start with the implementation we need to define what labels are.
+Labels are key-value pairs that can be used to filter and group metrics.
+Labels are used by monitoring systems like prometheus to create more dynamic dashboards.
+For example, a metric that counts the number of requests can have a label called `method` that defines if the request was a `GET` or `POST` request.
+By doing so the monitoring system can create a dashboard that shows the number of requests per method.
+
+In prometheus a concrete metric is defined like this (see https://prometheus.io/docs/concepts/data_model/#notation):
+
+```
+api_http_requests_total{method="POST", handler="/messages"}
+```
+
+Here a metric is defined by a name (`api_http_requests_total`) and a set of labels (`method="POST"` and `handler="/messages"`).
+In prometheus that defines the unique identifier of a metric.
+
+When using labels to create dashboards in Grafana prometheus provides a rich query language to filter and group metrics: https://prometheus.io/docs/prometheus/latest/querying/examples/
+
+We will use the same practices and definition as in prometheus regarding labels.
+
+### Public API
+
+The metrics module will be extended by a new data type called `Label`.
+A label is a key-value pair that can be defined as a `record`:
+
+```java
+record Label(@NonNull String key, @Nullable String value) {}
+``` 
+
+> [!NOTE]  
+The value of a label can be `null`.
+For our prometheus endpoint a label with a `null` value will be handled as if the label would not exist.
+
+The `Metric` interface will be extended by a new method called `getLabels()` that returns a list of labels:
+
+```java
+interface Metric {
+    
+    //...
+    
+    @NonNull
+    Set<Label> getLabels();
+}
+```
+
+Like the name or category, the labels are defined when a metric is created and cannot be changed later.
+Based on that the MetricConfig class will be extended by a new method called `withLabels`:
+
+```java
+abstract class MetricConfig {
+    
+    //...
+    
+    @NonNull
+    MetricConfig withLabels(@NonNull Label... labels) {...}
+    
+    @NonNull
+    MetricConfig withLabels(@NonNull Set<Label> labels) {...}
+}
+```
+
+By adding labels the unique identifier of a metric is the key and the labels.
+Based on that we need to modify the `Metrics` interface to use the labels as the unique identifier of a metric:
+
+```java
+
+interface Metrics {
+    
+    // old methods:
+    // Object getValue(@NonNull String category, @NonNull String name);
+    // void remove(@NonNull String category, @NonNull String name);
+    
+    @Nullable
+    Object getValue(@NonNull String category, @NonNull String name, @NonNull Set<Label> labels);
+    
+    void remove(@NonNull String category, @NonNull String name, @NonNull Set<Label> labels);
```

Comment:
if category and name are labels now, don't we need to simply supply the set of labels?